File: workers/research/src/black_book_research/discovery/ingestion.py
"""Bulk and API candidate ingestion interfaces (BB-039)."""

from __future__ import annotations

from dataclasses import dataclass

from black_book_research.adapters.candidates import assert_adapter_candidate_valid
from black_book_research.adapters.types import AdapterCandidateRecord
from black_book_research.query_packs.types import QueryPack

from .geography import extract_geographic_hints
from .identity import build_candidate_identity
from .signals import extract_discovery_signals
from .types import DISCOVERY_CANDIDATE_SCHEMA_VERSION, DiscoveryIngestMode, DiscoverySignal


@dataclass(frozen=True, slots=True)
class DiscoveryCandidateRecord:
    schema_version: str
    candidate_id: str
    identity_key: str
    stable_identifier: str
    ingest_mode: DiscoveryIngestMode
    signals: DiscoverySignal
    geographic_hint_count: int


def ingest_api_candidate(
    record: AdapterCandidateRecord,
    pack: QueryPack,
    *,
    candidate_id: str,
) -> DiscoveryCandidateRecord:
    assert_adapter_candidate_valid(record)
    identity = build_candidate_identity(record)
    signals = extract_discovery_signals(record, pack)
    hints = extract_geographic_hints(record)
    return DiscoveryCandidateRecord(
        schema_version=DISCOVERY_CANDIDATE_SCHEMA_VERSION,
        candidate_id=candidate_id,
        identity_key=identity.identity_key,
        stable_identifier=identity.stable_identifier,
        ingest_mode="api",
        signals=signals,
        geographic_hint_count=len(hints),
    )


def ingest_bulk_candidates(
    records: tuple[AdapterCandidateRecord, ...],
    pack: QueryPack,
    *,
    id_prefix: str = "disc",
) -> tuple[DiscoveryCandidateRecord, ...]:
    return tuple(
        ingest_api_candidate(
            record,
            pack,
            candidate_id=f"{id_prefix}_{index}_{record.stable_identifier}",
        )
        for index, record in enumerate(records)
    )

File: packages/domain/src/discovery/ingestion.ts
/**
 * Bulk and API candidate ingestion interfaces (BB-039).
 */
import {
  assertAdapterCandidateValid,
  validateAdapterCandidates,
  type AdapterCandidateRecord,
} from '../adapters/index.js';
import { buildCandidateIdentity } from './identity.js';
import { extractDiscoverySignals } from './signals.js';
import { extractGeographicHints } from './geography.js';
import {
  DISCOVERY_CANDIDATE_SCHEMA_VERSION,
  type ApiIngestRequest,
  type BulkIngestBatch,
  type DiscoveryCandidateRecord,
  type DiscoveryIngestMode,
} from './types.js';
import type { QueryPack } from '../query-packs/types.js';

export type IngestDiscoveryCandidateInput = {
  readonly record: AdapterCandidateRecord;
  readonly pack: QueryPack;
  readonly ingestMode: DiscoveryIngestMode;
  readonly candidateId: string;
  readonly now: string;
};

function buildDiscoveryCandidate(input: IngestDiscoveryCandidateInput): DiscoveryCandidateRecord {
  assertAdapterCandidateValid(input.record);
  return {
    schemaVersion: DISCOVERY_CANDIDATE_SCHEMA_VERSION,
    id: input.candidateId,
    identity: buildCandidateIdentity(input.record),
    adapterRecord: input.record,
    status: 'pending',
    ingestMode: input.ingestMode,
    signals: extractDiscoverySignals(input.record, input.pack),
    geographicHints: extractGeographicHints(input.record),
    retryCount: 0,
    createdAt: input.now,
    updatedAt: input.now,
  };
}

/** Bulk ingestion interface for adapter batch output. */
export function ingestBulkCandidates(
  batch: BulkIngestBatch,
  pack: QueryPack,
  options: { readonly now: string; readonly idPrefix?: string },
): readonly DiscoveryCandidateRecord[] {
  validateAdapterCandidates(batch.records);
  const mode = batch.ingestMode ?? 'bulk';
  const prefix = options.idPrefix ?? 'disc';

  return batch.records.map((record, index) =>
    buildDiscoveryCandidate({
      record,
      pack,
      ingestMode: mode,
      candidateId: `${prefix}_${index}_${record.stableIdentifier}`,
      now: options.now,
    }),
  );
}

/** API ingestion interface for single candidate records. */
export function ingestApiCandidate(
  request: ApiIngestRequest,
  pack: QueryPack,
  options: { readonly now: string; readonly candidateId?: string },
): DiscoveryCandidateRecord {
  return buildDiscoveryCandidate({
    record: request.record,
    pack,
    ingestMode: 'api',
    candidateId: options.candidateId ?? `disc_api_${request.record.stableIdentifier}`,
    now: options.now,
  });
}

export type BulkIngestInterface = typeof ingestBulkCandidates;
export type ApiIngestInterface = typeof ingestApiCandidate;
